crates/module-system/module-implementations/sov-test-state-consistency/src/call.rs

```diff
@@ -0,0 +1,162 @@
+    Debug,
+    PartialEq,
+    Eq,
+    Clone,
+    JsonSchema,
+    EnumDiscriminants,
+    EnumIs,
+    UniversalWallet,
+)]
+#[serde(rename_all = "snake_case")]
+#[strum_discriminants(derive(VariantArray, EnumIs))]
+pub enum CallMessage {
+    /// Insert a new value into the state, asserting the old one matches the current storage.
+    UpdateValue {
+        /// The current value in the module. Asserted to ensure state consistency; the transaction
+        /// is invalid on mismatch.
+        old_check: u64,
+        /// The new value to save.
+        new: u64,
+    },
+    /// Updates the value stored in accessory state.
+    UpdateAccessoryState(u64),
+    /// Assert the state accessor's block and slot properties are as expected.
+    AssertBlockState {
+        /// The expected visible slot number.
+        expected_visible_slot_number: u64,
+        /// The expected rollup height.
+        expected_rollup_height: u64,
+        /// The expected state root.
+        expected_state_root: Vec<u8>,
+    },
+}
+
+impl<S: Spec> StateConsistency<S> {
+    pub(crate) fn update_value(
+        &mut self,
+        old_check: u64,
+        new: u64,
+        context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        // Get the current value for this sender's address
+        let sender_address = context.sender();
+        let current_value = self.values.get(sender_address, state)?.unwrap_or(0);
+
+        anyhow::ensure!(
+            current_value == old_check,
+            "Current value mismatched: stored value {}, transaction expected {}",
+            current_value,
+            old_check
+        );
+
+        // Update the value for this address
+        self.values.set(sender_address, &new, state)?;
+
+        // Emit event
+        self.emit_event(
+            state,
+            Event::ValueUpdated {
+                address: sender_address.clone(),
+                old_value: current_value,
+                new_value: new,
+            },
+        );
+
+        Ok(())
+    }
+
+    pub(crate) fn update_accessory_state(
+        &mut self,
+        new: u64,
+        _context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        Ok(self.accessory_value.set(&new, state)?)
+    }
+
+    pub(crate) fn assert_block_state(
+        &mut self,
+        expected_visible_slot_number: u64,
+        expected_rollup_height: u64,
+        expected_state_root: Vec<u8>,
+        _context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        let mut mismatches = Vec::<String>::new();
+
+        let rollup_height = state.rollup_height_to_access().get();
+        // Sanity check
+        let saved_rollup_height = self.latest_rollup_height.get(state)?.unwrap_or(0);
+        assert_eq!(rollup_height, saved_rollup_height, "Mismatch between rollup height saved from begin_rollup_block_hook: {saved_rollup_height} and the one read from TxState: {rollup_height}");
```

Comment:
Here we panic on failure. Is that really the behavior we want? Should this return an error instead?